File: Ruby/fib.rb
#ruby --jit-wait fib.rb

def fib_recursive(n)
  	return n if n <= 1
  	fib_recursive(n - 1) + fib_recursive(n - 2)
end

def fib_iterative(n)
	a = 0
	b = 1
	(0...n).each {
		c = a + b
		a, b = b, c
	}
	a
end

# def prompt(*args)
#     print(*args)
#     gets
# end

start_time = Time.now
puts fib_recursive(42)
elapsed_time = Time.now - start_time
printf("Seconds Recursive took: %5.3f s.\n", elapsed_time)

# n = (prompt "Fib number: ").to_i

start_time = Time.now
puts fib_iterative(42)
elapsed_time = Time.now - start_time
printf("Seconds Iterative took: %5.3f s.\n", elapsed_time)

File: C/fib.c
// clang -O3 fib.c -o fib

#include <stdio.h>
#include <stdint.h>
#import <time.h>

uint64_t fib_recursive(uint64_t n) {
	if (n <= 1) return n;
	return fib_recursive(n - 1) + fib_recursive(n - 2);
}

uint64_t fib_iterative(uint64_t n) {
	uint64_t a = 0;
	uint64_t b = 1;
	for (int i = 0; i < n; i++) {
		uint64_t c = a + b;
		a = b;
		b = c;
	}
	return a;
}

int main(void) {
	clock_t     start, end;
	double      r;

	start = clock(); // start the clock
	printf("%llu \n", fib_recursive(42));
	end = clock(); // stop the clock
    r = ((double)(end-start)) / ((double)(CLOCKS_PER_SEC));
    printf("Recursive took %.3f Seconds.\n", r);

    start = clock(); // start the clock
	printf("%llu \n", fib_iterative(42));
	end = clock(); // stop the clock
    r = ((double)(end-start)) / ((double)(CLOCKS_PER_SEC));
    printf("Iterative took %.3f Seconds.\n", r);

	return 0;
}